### DISCUSSION TITLE
MikroORM bulk actions - write and read, use entities as adapters?

### DISCUSSION THREAD

**Discussion by DecathectZero** [mikro-orm/mikro-orm#5272] [Q&A]
We currently have an application that loads in projects for users. Each project may contain 50k+ rows of a particular table. We also need to be able to duplicate projects, so that is a bulk write of 50k rows.
We find that using Mikro entities tanks the performance, if every single row gets converted to an entity, the memory and compute overhead is immense. 

The memory overhead is immense on Mikro, we had to build a parallel system that reads/writes the JSON from the DB and a whole separate set of adapters to match the `entity.toJSON`.
![image](https://github.com/mikro-orm/mikro-orm/assets/5643574/dfe9071a-8e31-4228-a834-b43cb26ec6e8)

Is there some way to just use the entity as a lightweight wrapper/adapter to transform raw DB JSON to properties defined in the entity? For example it is `environmentID` in the entity but `environment_id` in the DB.

**Comment by B4nan** [comment]
Maybe we could optimize this, can you share a complete repro?

>  For example it is environmentID in the entity but environment_id in the DB.

This is handled on the result set mapper level, you don't have to map to entities for that, but you'd need to use the QB instead of the EM methods, namely `qb.execute()` which by defaults maps the field names (and also merges joined results to an entity graph, in case you join to-many relations). Mapping to entities is handled level above in the hydration layer, which is what you get with the EM methods or by awaiting the QB (which calls `qb.getResultList()`).

**Reply by DecathectZero** [reply]
We've noted that `qb.execute()` is also quite slow as well, we believe there's a long of additional logic around caching right?

https://github.com/mikro-orm/mikro-orm/blob/27f0c83e11ea74513279c05b84c81707a1e7e8c3/packages/knex/src/query/QueryBuilder.ts#L740-L795

**Reply by B4nan** [reply]
It should be pretty fast, result cache is opt-in, so that shouldn't be relevant.

Hard to reason about this since I don't know what you are doing, how complex your entities/tables are, etc. If you want to talk perf issues, I need to see a complete repro first.

**Reply by DecathectZero** [reply]
Appreciate the fast response! We'll need some time to put together the repro, since a lot of it is very coupled to our codebase. We'll try to make some more isolated benchmarks.

**Reply by z4o4z** [reply]
hey @B4nan here’s an example of our entities:
PostgresCMSObjectEntity, PostgresCMSTabularEntity are abstract classes with a few additional fields like id, updatedAt, updatedBy, assistant (reference to another entity), and etc.

```ts
@Entity({ tableName: 'designer.utterance' })
@Unique({ properties: ['id', 'environmentID'] })
@Index({ properties: ['environmentID'] })
export class UtteranceEntity extends PostgresCMSObjectEntity {
  static fromJSON<JSON extends Partial<ToJSONWithForeignKeys<UtteranceEntity>>>(data: JSON) {
    return UtteranceEntityAdapter.toDB<JSON>(data);
  }

  @Property({ type: UtteranceTextType })
  text: UtteranceText;

  @ManyToOne(() => IntentEntity, {
    name: 'intent_id',
    onDelete: 'cascade',
    fieldNames: ['intent_id', 'environment_id'],
  })
  intent: Ref<IntentEntity>;

  @Enum(() => Language)
  language: Language;

  @Assistant()
  assistant: Ref<AssistantEntity>;

  @Environment()
  environmentID: string;

  [PrimaryKeyType]?: CMSCompositePK;

  constructor(data: EntityCreateParams<UtteranceEntity>) {
    super(data);

    ({
      text: this.text,
      intent: this.intent,
      language: this.language,
      assistant: this.assistant,
      environmentID: this.environmentID,
    } = UtteranceEntity.fromJSON(data));
  }

  toJSON(...args: any[]): ToJSONWithForeignKeys<UtteranceEntity> {
    return UtteranceEntityAdapter.fromDB({
      ...wrap<UtteranceEntity>(this).toObject(...args),
      intent: this.intent,
      updatedBy: this.updatedBy,
      assistant: this.assistant,
    });
  }
}

@EntityDecorator({ tableName: 'designer.required_entity' })
@Index({ properties: ['environmentID'] })
export class RequiredEntityEntity extends PostgresCMSObjectEntity {
  static fromJSON<JSON extends Partial<ToJSONWithForeignKeys<RequiredEntityEntity>>>(data: JSON) {
    return RequiredEntityEntityAdapter.toDB<JSON>(data);
  }

  @ManyToOne(() => EntityEntity, {
    name: 'entity_id',
    onDelete: 'cascade',
    fieldNames: ['entity_id', 'environment_id'],
  })
  entity: Ref<EntityEntity>;

  @ManyToOne(() => IntentEntity, {
    name: 'intent_id',
    onDelete: 'cascade',
    fieldNames: ['intent_id', 'environment_id'],
  })
  intent: Ref<IntentEntity>;

  @ManyToOne(() => ResponseEntity, {
    name: 'reprompt_id',
    default: null,
    onDelete: 'set default',
    nullable: true,
    fieldNames: ['reprompt_id', 'environment_id'],
  })
  reprompt: Ref<ResponseEntity> | null = null;

  @Assistant()
  assistant: Ref<AssistantEntity>;

  @Environment()
  environmentID: string;

  [PrimaryKeyType]?: CMSCompositePK;

  constructor(data: EntityCreateParams<RequiredEntityEntity>) {
    super(data);

    ({
      entity: this.entity,
      intent: this.intent,
      reprompt: this.reprompt,
      assistant: this.assistant,
      environmentID: this.environmentID,
    } = RequiredEntityEntity.fromJSON(data));
  }

  toJSON(...args: any[]): ToJSONWithForeignKeys<RequiredEntityEntity> {
    return RequiredEntityEntityAdapter.fromDB({
      ...wrap<RequiredEntityEntity>(this).toObject(...args),
      entity: this.entity,
      intent: this.intent,
      reprompt: this.reprompt ?? null,
      updatedBy: this.updatedBy,
      assistant: this.assistant,
    });
  }
}

@Entity({ tableName: 'designer.intent' })
@Unique({ properties: ['id', 'environmentID'] })
@Index({ properties: ['environmentID'] })
export class IntentEntity extends PostgresCMSTabularEntity {
  static fromJSON<JSON extends Partial<ToJSONWithForeignKeys<IntentEntity>>>(data: JSON) {
    return IntentEntityAdapter.toDB<JSON>(data);
  }

  // legacy built-in intents uses type as id, so increase length to 64
  @PrimaryKey({ type: 'varchar', nullable: false, length: 64 })
  id!: string;

  // to keep composite key correct, environmentID must be defined after id
  @Environment()
  environmentID!: string;

  @Property({ type: 'text', default: null, nullable: true })
  description: string | null;

  @Property()
  automaticReprompt: boolean;

  @Property({ type: ArrayType })
  entityOrder: string[];

  @OneToMany('UtteranceEntity', (value: UtteranceEntity) => value.intent)
  utterances = new Collection<UtteranceEntity>(this);

  @OneToMany('RequiredEntityEntity', (value: RequiredEntityEntity) => value.intent)
  requiredEntities = new Collection<RequiredEntityEntity>(this);

  constructor({ description, entityOrder, automaticReprompt, ...data }: EntityCreateParams<IntentEntity>) {
    super(data);

    ({
      description: this.description,
      entityOrder: this.entityOrder,
      automaticReprompt: this.automaticReprompt,
    } = IntentEntity.fromJSON({ description, entityOrder, automaticReprompt }));
  }

  toJSON(...args: any[]): ToJSONWithForeignKeys<IntentEntity> {
    return IntentEntityAdapter.fromDB({
      ...wrap<IntentEntity>(this).toObject(...args),
      folder: this.folder ?? null,
      updatedBy: this.updatedBy,
      createdBy: this.createdBy,
      assistant: this.assistant,
    });
  }
}
```

the query we run for all entities is pretty simple, no joins and etc:

```ts
 findManyByEnvironment(assistant: PKOrEntity<AssistantEntity>, environmentID: string): Promise<Entity[]> {
      return this.find(
        { assistant, environmentID } as FilterQuery<Entity>,
        { orderBy: { createdAt: 'DESC' } } as FindOptions<Entity>
      );
    }
```

**Reply by z4o4z** [reply]
@B4nan I've dived deeper into the perf issue and found that entities creation takes most of the time. Here's the logs from the `.find` method

<img width="233" alt="image" src="https://github.com/mikro-orm/mikro-orm/assets/9130301/8c505e79-6aba-4542-aa71-e79d34054be7">

**Reply by B4nan** [reply]
I'd appreciate a complete reproduction - a script I can run and see what you see, including creating the data.

Also, have you tried `qb.execute` as suggested? That way there won't be any hydration. I'd expect the slowdown being caused by propagation support, as all the properties are defined as get/set pair to allow change detection.

**Reply by z4o4z** [reply]
yeah, I've just tried the execute method, it works way faster but it returns data in other formats (db one, or row entity) so we need to write custom adapters. And another problem we need to use the `.upsertMany` method which has the same perf issue as `.find`. 

I'll try to build a script, but it'll take some time. Thanks for helping us with it :)

**Reply by B4nan** [reply]
You can use QB for upserting too (via `qb.onConflict()`).

**Reply by B4nan** [reply]
>     onDelete: 'set default',

Seeing this, you are still on v5, right? If we find a way to make things faster, it will be in v6, I don't want to make bigger changes to 5.x branch, only smaller bug fixes that are easily portable can get there.

**Reply by z4o4z** [reply]
@B4nan here's the example https://github.com/z4o4z/mikro-perf-reproduction, it shows that `.find` is ~8x slower then `.qb().execute()`

**Reply by z4o4z** [reply]
yeah sure, we have plans to update to the latest version if the issue is fixed

**Reply by B4nan** [reply]
Thanks, will try to have a look over the weekend to see what's happening. With that said, I was hoping for something much smaller than this, its very complex setup with a lot of custom abstraction, and first thing will be to greatly simplify it. I am really looking for a single file repro, with literally no abstraction, and simplifying this might easily take a few hours on its own. 

Also, why the `Promise.all` calls, are those significant for anything?

**Reply by B4nan** [reply]
Gave it an initial look, a few findings:
- a significant part of the time is produced by your JSON type extension (`UtteranceTextType`) which validates data via zod, on my end it's around 25% (more than 500ms)
- you are missing `ref: true` in your to-one properties that use `Ref` wrapper (so your entities hydrated from database wont use the `Ref` wrapper)
- the idea of accepting raw db data in your entity constructors seems a bit wrong to me, or at least it goes against how the ORM is designed, and if you would use some methods that create new entity instances (e.g. `em.create` but also simple `em.assign`), such constructors would get different parameters (always property names, not column names, and for custom types the values are always converted to runtime ones)
- the fact that you use composite PKs has quite a significant effect, those will be always harder to map

I managed to do a few smaller improvements and shaved off another 150ms, so I am now at ~1150ms from the initial ~1900ms (when using regular json mapping instead of that zod parsing as mentioned above). I am sure there are more small places that can be optimized further, but its a detective work, there doesn't seem to be one huge bottleneck, its quite spread - the hydration takes most of time, but even without that its still around 800ms, with almost 200ms taking the registration to the identity map (`uow.register`).

**Reply by B4nan** [reply]
I've identified a few smaller bottlenecks in how composite PKs are handled when checking identity map, refactoring this won't be that easy, but it should also help with almost 200ms (down to ~950ms now, but the ORM test suite is not completely green with those changes). I don't see any room for improvement in the hydration though, but maybe it will be more about similar unrelated changes that will help with that too.

edit: ok, it wasn't that hard: 73094efb8f140f57bf6e1f160edf209812182eab (although the effect is not that significant in the end, I am around 980ms now)

btw I guess I don't need a simpler repro right now, as the issue is in loading entities, its not connected to your additional abstractions in the constructors etc
